File: task1.cpp
#include "task1.h"

void readfile(string infile, string* lines) {
	ifstream fin(infile);
	fin >> lines[0] >> lines[1];
	fin.close();
};

int EditDis(string a, string b, int *Ops, char *Obj, int *Pos) {
	int m = a.length();
	int n = b.length();
	int **edit, **path;
	edit = new int*[m + 1];
	path = new int*[m + 1];
	for (int i = 0; i <= m; i++) {
		edit[i] = new int[n + 1]{ 0 };
		path[i] = new int[n + 1]{ 0 };
	}
	for (int i = 0; i <= m; i++)
		edit[i][0] = i;
	for (int j = 0; j <= n; j++)
		edit[0][j] = j;
	for (int i = 1; i <= m; i++) {
		for (int j = 1; j <= n; j++) {
			int diff = (a[i - 1] == b[j - 1] ? 0 : 1);
			int DEL = edit[i - 1][j] + 1;
			int INS = edit[i][j - 1] + 1;
			int SUB = edit[i - 1][j - 1] + diff;

			int op;
			if (DEL <= INS && DEL <= SUB) {
				edit[i][j] = DEL;
				op = 0;
			}
			else if (INS <= DEL && INS <= SUB) {
				edit[i][j] = INS;
				op = 1;
			}
			else {
				edit[i][j] = SUB;
				op = 2;
			}

			if (op == 2 && diff == 0)
				path[i][j] = -1;
			else
				path[i][j] = op;
		}
	}
	int dist = edit[m][n];
	int i = m;
	int j = n;
	int cnt = 0;
	while (i != 0 || j != 0) {
		switch (path[i][j]) {
		case 0:
			Ops[cnt] = 0;
			Obj[cnt] = '-';
			Pos[cnt] = i - 1;
			cnt++;
			i--;
			break;
		case 1:
			Ops[cnt] = 1;
			Obj[cnt] = b[j - 1];
			Pos[cnt] = i;
			cnt++;
			j--;
			break;
		case 2:
			Ops[cnt] = 2;
			Obj[cnt] = b[j - 1];
			Pos[cnt] = i - 1;
			cnt++;
			i--;
			j--;
			break;
		default:
			i--;
			j--;
			break;
		}
	}
	return dist;
}

void writefile(string outfile, int dist, int *Ops, char *Obj, int *Pos) {
	ofstream fout(outfile);
	fout << dist << '\n';
	for (int i = dist - 1; i >= 0; i--) {
		int op;
		string outstr;
		op = Ops[i];
		if (op == 0 || op == -1)
			outstr = Operate[op] + ' ' + to_string(Pos[i]);
		else
			outstr = Operate[op] + ' ' + to_string(Pos[i]) + ' ' + Obj[i];
		fout << outstr << '\n';
	}
	fout.close();
}

void Task1(string infile, string outfile) {
	string *lines = new string[2];
	readfile(infile, lines);
	string a = lines[0];
	string b = lines[1];
	int dist;
	int *Ops, *Pos;
	char *Obj;
	Ops = new int[10000];
	Pos = new int[10000];
	Obj = new char[10000];
	dist = EditDis(a, b, Ops, Obj, Pos);
	writefile(outfile, dist, Ops, Obj, Pos);
}

File: Task1/Python/task1.py
import argparse
import types
import time
from tools import *

Operate = ["DEL", "INS", "SUB"]

# Read standard File
# Input: 
	# filename: str
# Output: lines
	# lines: list of str
def readfile(filename):
	file = open(filename, 'r')
	lines = file.readlines()
	lines = [x.strip() for x in lines]
	lines = [x for x in lines if x]
	file.close()
	return lines

# Calculate the Edit Distance
# Input:
	# a, b: str, a to b
# Output:
	# dist: int, the edit distance of a to b
	# Ops: list of int, Operate idx
		# if Op == 1 or 2, the element will be (2, 'A')
	# Pos: list of int, position of a
def EditDis(a, b):
	m = len(a)
	n = len(b)
	edit = [[0 for col in range(n+1)] for row in range(m+1)]
	path = [[0 for col in range(n+1)] for row in range(m+1)]
	# Ops = [[0 for col in range(n)] for row in range(m)]
	# Pos = [[0 for col in range(n)] for row in range(m)]
	Ops = []
	Pos = []
	for i in range(m+1):
		edit[i][0] = i
	for j in range(n+1):
		edit[0][j] = j
	for i in range(1, m+1):
		for j in range(1, n+1):
			diff = 0 if (a[i-1] == b[j-1]) else 1
			DEL = edit[i-1][j] + 1
			INS = edit[i][j-1] + 1
			SUB = edit[i-1][j-1] + diff
			temp = [DEL, INS, SUB]
			edit[i][j] = min(temp)
			op = temp.index(min(temp))
			if op == 2 and diff == 0:
				path[i][j] = -1
			else:
				path[i][j] = op
	dist = edit[m][n]
	# RePath(a, b, path, m, n, Ops, Pos)
	i = m
	j = n
	while(i != 0 or j != 0):
		if path[i][j] == -1:
			i = i - 1
			j = j - 1
		elif path[i][j] == 0: # DEL
			Ops.append(0)
			Pos.append(i-1)
			i = i - 1
		elif path[i][j] == 1: # INS
			Ops.append((1, b[j-1]))
			Pos.append(i)
			j = j - 1
		elif path[i][j] == 2: # SUB
			Ops.append((2, b[j-1]))
			Pos.append(i-1)
			i = i - 1
			j = j - 1
	print edit
	return dist, Ops, Pos	

# Write standard File
# Input:
	# outfile: str
	# dist: int, the edit distance of a to b
	# Ops: list of int, Operate idx
	# if Op == 1 or 2, the element will be (2, 'A')
	# Pos: list of int, position of a
# Output:
	# None
def writefile(outfile, dist, Ops, Pos):
	out = open(outfile, 'w')
	out.write(str(dist) + '\n')
	# print(str(dist) + '\n')
	for i in range(dist-1, -1, -1):
		op = Ops[i]
		if type(op) is types.IntType:
			outstr = Operate[op] + ' ' + str(Pos[i]) + '\n'
		else:
			outstr = Operate[op[0]] + ' ' + str(Pos[i]) +  ' ' + op[1] + '\n'
		# print(outstr)
		out.write(outstr)
	out.close()
	

	
# Task 1: read infile and write result to outfile
# Input:
	# infile: str
	# outfile: str
# Output:
	# None
def Task1(infile, outfile):
	lines = readfile(infile)
	if len(lines) != 2:
		print("Input lines overflow!")
		return
	a = lines[0]
	b = lines[1]
	[dist, Ops, Pos] = EditDis(a,b)
	writefile(outfile, dist, Ops, Pos)
	
if __name__ == '__main__':
	parser = argparse.ArgumentParser()
	parser.add_argument('-i', metavar='infile')
	parser.add_argument('-o', metavar='outfile')
	args = parser.parse_args()

	start = time.clock()
	Task1(args.i, args.o)
	end = time.clock()
	
	print("Running time is : " + str(end-start))

# a = 'ABCDABADBDC'
# b = 'BCDBBCC'
# a = 'kitten'
# b = 'sitting'
# [dist, Ops, Pos] = EditDis(a,b)
# writefile('out.txt', dist, Ops, Pos)

	